splice_machine/src/main/java/com/splicemachine/derby/impl/sql/execute/operations/window/function/FirstLastValueFunction.java

```diff
@@ -60,7 +60,37 @@ protected void calculateOnAdd(WindowChunk chunk, DataValueDescriptor[] dvds) thr
 
     @Override
     protected void calculateOnRemove(WindowChunk chunk, DataValueDescriptor[] dvds) throws StandardException {
-       // nothing to do here we've already set all the results. For LAST_VALUE(), result is just the last item we added.
+       if (isLastValue) {
+           // we need to handle the scenario where the last value is the one that is removing, this could happen
+           // when ignoreNulls is true
+           DataValueDescriptor removedVal = dvds[0];
+           if (ignoreNulls && !removedVal.isNull() && chunk.getResult().equals(removedVal)) {
+               int position = chunk.last;
+               DataValueDescriptor result = null;
+               while (position > chunk.first) {
+                   result = chunk.get(position-1)[0];
+                   if (result !=null && !result.isNull()) {
+                       chunk.setResult(result);
+                       return;
+                   }
+                   position --;
+               }
+               chunk.setResult(result);
+           }
+       } else {// handle First_Value
+           int position = chunk.first;
+           DataValueDescriptor result = null;
+           while (position < chunk.last) {
+               result = chunk.get(position)[0];
+               if (result !=null && (!ignoreNulls || !result.isNull())) {
```

Comment:
The loop to handle Last_Value decrements all the way back to chunk.first, because it uses 'position-1' as the index passed to chunk.get(), but the First_Value loop only iterates up to chunk.last-1, so doesn't include the last value in the window.  Is this by design, or should the while loop really use 'while (position **<=** chunk.last) {'   ?   

It's not clear if the value being removed is currently included in between chunk.first and chunk.last, or if the window has already been updated by the time we get in here, and the removed value is not in this range.  Knowing this might help answer if chunk.last should be included.

Reply:
The valid values in the chunk are from position first to last-1. 
The value being removed is not included in chunk.first and chunk.last-1. It may be clearer to look from the caller of calculateOnRemove() in SpliceGenericWindowFunction.remove().